timer/timer.go

```diff
@@ -0,0 +1,144 @@
+package timer
+
+import (
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// Timer definition
+type (
+	Timer struct {
+		ID             uuid.UUID
+		Domain         string
+		ExecutionCount uint32
+		Meta           Meta
+		Task           Task
+		Schedule       Schedule
+	}
+
+	// Task represents a timer task.
+	Task interface {
+		isTask()
+	}
+
+	// Schedule represents the "when" configuration of a timer.
+	Schedule interface {
+		isSchedule()
+	}
+
+	// Meta represents timer metadata.
+	Meta struct {
+		CreationTime time.Time
+	}
+)
+
+// HTTP Task represents an HTTP request.
+type (
+	HTTP struct {
+		URL string
+		Method
+		Body    string
+		Headers map[string]string
+	}
+
+	// Method is the method of the HTTP request.
+	Method int
+)
+
+func (HTTP) isTask() {}
+
+// Method definitions
+const (
+	GET Method = iota
+	POST
+)
+
+// Schedule definitions
+type (
+	// Cron is a schedule configured by a cron string.
+	Cron struct {
+		Start      time.Time
+		Cron       string
+		Executions int32
+	}
+
+	// Interval is a schedule configured by a fixed interval.
+	Interval struct {
+		Start      time.Time
+		Interval   int32
+		Executions int32
+	}
+)
+
+func (Cron) isSchedule() {}
+
+func (Interval) isSchedule() {}
+
+// MarshalBinary marshals a timer to a []byte.
+func (t *Timer) MarshalBinary() ([]byte, error) {
+	// Use json package to turn timer into map[string]interface
+	bytes, err := json.Marshal(t)
+	if err != nil {
+		return nil, err
+	}
+	var m map[string]interface{}
+	err = json.Unmarshal(bytes, &m)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set "type indicators"
+	var sched string
+	switch t.Schedule.(type) {
+	case *Cron:
+		sched = "cron"
+	case *Interval:
+		sched = "interval"
+	default:
+		panic("Can't marshal schedule, unknown type.")
+	}
+	m["scheduleType"] = sched
```

Comment:
This is weird to marshal something with an interface and then add the interface back in. You probably just want your types to be serializable (I would actually just start with proto for these as well).

Reply:
I don't understand this. Timers have to be marshalled and they also semantically have interface fields. Are you saying I should just have protobufs as an intermediate type? I've done that and it works, but the conversion from proto type to regular type is just as nasty, but ya I can do it. Or are you saying timers shouldn't have interface fields, in which case I'd like to understand what that would look like.